reconciler/testing/context.go

```diff
@@ -43,3 +53,74 @@ func SetupFakeContextWithCancel(t zaptest.TestingT) (context.Context, context.Ca
 	ctx, is := injection.Fake.SetupInformers(ctx, &rest.Config{})
 	return ctx, c, is
 }
+
+// fakeClient is an interface capturing the two functions we need from
+type fakeClient interface {
+	PrependWatchReactor(resource string, reaction clientgotesting.WatchReactionFunc)
+	PrependReactor(verb, resource string, reaction clientgotesting.ReactionFunc)
+}
+
+// withTracker is an interface capturing only the Tracker function. The dynamic client
+// currently does not have that, so we need to special-case it.
+type withTracker interface {
+	Tracker() clientgotesting.ObjectTracker
+}
+
+// RunAndSyncInformers runs the given informers, then makes sure their caches aer all
+// synced and in addition makes sure that all the Watch calls have been properly setup.
+// See https://github.com/kubernetes/kubernetes/issues/95372 for background on the Watch
+// calls tragedy.
+func RunAndSyncInformers(ctx context.Context, informers ...controller.Informer) (func(), error) {
+	var watchesPending atomic.Int32
```

Comment:
out of interest, why atomics and not waitGroup?

Reply:
you can't call Add and Done in parallel.

Reply:
really? I dont see that in the docs, am I missing it? (I think Done is basically just Add(-1))

Reply:
Actually, you can. You must call all `Add` before calling `Wait` though. I *think* that invariant might be true here because all the `Add` would be in the List call which should finish before the sync is finished. But I'm a bit unsure on that front so I'd leave it like this.

This is also taken from the example code which specically comments on not using a WaitGroup: https://github.com/istio/istio/blob/d346433aa3fc1315f9bbd8e413340941a60f6c89/pkg/kube/client.go#L231

Reply:
ah, add and _wait_, gotcha. ok